File: Python/Medium/graphs/clone_graph.py
class Node(object):
    def __init__(self, val=0, neighbors = []):
        self.val = val
        self.neighbors = neighbors


class Solution:
    def cloneGraph(self, node):
        if not node:
            return None
        
        clone_dict = {}
        queue = []

        rootClone = Node(node.val, [])
        clone_dict[node.val] = rootClone
        queue.append(node)

        while queue:
            cur = queue.pop()
            curClone = clone_dict[cur.val]

            for neighbor in cur.neighbors:
                if neighbor.val not in clone_dict:
                    newClone = Node(neighbor.val, [])
                    clone_dict[neighbor.val] = newClone
                    queue.append(neighbor)

            curClone.neighbors.append(clone_dict[neighbor.val]) 

        return rootClone

File: JavaScript/Medium/graphs/clone_graph.js
function Node(val, neighbors) {
    this.val === undefined ? 0 : val;
    this.neighbors === undefined ? [] : neighbors;
}

/**
 * @param {Node}
 * @return {Node}
 */

var cloneGraph = function(node) {
    if (!node) {
        return null
    }

    let cloneDict = {}
    let q = []

    root = new Node(node.val, [])
    cloneDict[node.val] = root
    q.push(node)

    while (q.length) {
        cur = q.pop()
        curClone = cloneDict[cur.val]

        for (let neighbor of cur.neighbors) {
            if (!cloneDict[neighbor.val]) {
                newClone = new Node(neighbor.val, [])
                cloneDict[neighbor.val] = newClone
                q.push(neighbor)
            }

            curClone.neighbors.push(cloneDict[neighbor.val])
        }
    }

    return root

}